### PR TITLE
Implement bandstructure model by Massacrier et al

### PR DESCRIPTION
**Pull request by timcallow**
This PR implements the bandstructure model for AA codes suggested by [Massacrier](https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.023026) et al.

This is implemented by introducing an extra boundary condition, "bands". To facilitate the implementation of this extra model, and make the code overall more readable, the old `lbound`, `lunbound` and `occnums_bound`, `occnums_unbound` properties of `staticKS.Orbitals` have been merged into the single properties `ldegen` and `occnums`. Furthermore, the `occnums` parameter is now just pure Fermi-Dirac occupation numbers, and `occnums_w` denotes the occupation weighted with degeneracies and the new DOS parameter.

Various properties of the `staticKS.Orbitals` class now have an extra dimension to account for the number of "k points" in each band (always equal to 1 for the existing "neumann" and "dirichlet" bcs, and is chosen by the user for the new "bands" bc).

### CODE REVIEW

**Comment by timcallow** [comment]
@DanielKotik This is ready for you to take a look at. Sorry it's a bit long!

**Comment by DanielKotik** [comment]
I'll try to have a look today!

File: atoMEC/writeoutput.py [hunk 1]
```diff
@@ -396,11 +408,15 @@ def write_final_energies(energy):
         ent = energy.entropy
         ent_str = "{S:30s} : {S_x:10.4f}".format(S="Entropy", S_x=ent["tot"]) + spc
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:409]: I find this old style of f-strings rather unreadable. But we had already discussed this. Would be good, to switch at some point to the new style (best would be a dedicated PR) ...

File: atoMEC/writeoutput.py [hunk 2]
```diff
@@ -510,16 +525,16 @@ def density_to_csv(rgrid, density, filename):
...
-            + "n^dw_b"
-            + 4 * " "
-            + "n^dw_ub"
+            + "n^u (ideal)"
+            + 2 * " "
+            + "n^d (orbs)"
             + 3 * " "
+            + "n^d (ideal)"
+            + 1 * " "
         )
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:538]: I do not understand your intention here. Why is it necessary to assemble the string from small parts? And why so complicated?
Suggestion, something like
```python
headstr = r" ".join(["r (a_0)", "n^u (orbs)", "n^u (ideal)", "n^d (orbs)", "n^d (ideal)"])
```

File: atoMEC/writeoutput.py [hunk 3]
```diff
@@ -531,12 +546,12 @@ def density_to_csv(rgrid, density, filename):
             ]
         )
     else:
-        headstr = "r" + 8 * " " + "n_b" + 6 * " " + "n^_ub" + 3 * " "
+        headstr = "r (a_0)" + 4 * " " + "n (orbs)" + 5 * " " + "n (ideal)" + 3 * " "
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:549]: See comment above.

File: atoMEC/writeoutput.py [hunk 4]
```diff
@@ -587,6 +602,101 @@ def potential_to_csv(rgrid, potential, filename):
...
+        + 5 * " "
+        + "occs"
+        + 6 * " "
+        + "dos"
+        + 7 * " "
+        + "l_degen"
+        + 3 * " "
+        + "k_int_wt"
+        + 4 * " "
+    )
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:649]: See comment above.

File: atoMEC/writeoutput.py [hunk 5]
```diff
@@ -587,6 +602,101 @@ def potential_to_csv(rgrid, potential, filename):
...
+        data = np.column_stack([e_arr[:, sp], fd_arr[:, sp], DOS_arr[:, sp]])
+
+        try:
+            data_tot = np.concatenate((data_tot, data), axis=-1)
+        except ValueError:
+            data_tot = data
+
+    headstr = config.spindims * (
+        "energy" + 5 * " " + "fd occ" + 6 * " " + "dos" + 10 * " "
+    )
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:688]: See comment above.

File: atoMEC/numerov.py [hunk 6]
```diff
@@ -579,9 +581,262 @@ def num_propagate(xgrid, v, l, E):
...
+#         ) / (1.0 + h * W[i])
+
+#     # normalize the wavefunction
+#     Psi = Psi.transpose()
+#     psi_sq = np.exp(-xgrid) * Psi ** 2  # convert from P_nl to X_nl and square
+#     integrand = 4.0 * np.pi * np.exp(3.0 * xgrid) * psi_sq
+#     norm = (np.trapz(integrand, x=xgrid, axis=-1)) ** (-0.5)
+#     Psi_norm = np.einsum("ijl,ijlk->ijlk", norm, Psi)
+
+#     return Psi_norm
```
Comment (DanielKotik) [inline atoMEC/numerov.py:842]: Mhh, I don't really like the idea  of having a bunch of block comments of old code in the code base. My suggestion would be to create a special commit with a clear message indicating the removal of deprecated code that may become useful again in future versions. Also what is not clear to me is where this code is coming from? It doesn't look like you are converting the code to a block comment, but adding commented code here :thinking:.

**Review by DanielKotik — CHANGES_REQUESTED** [review]
I've added a few comments, but nothing spectacular :)

File: atoMEC/writeoutput.py (hunk 2 above)
Comment (timcallow) [reply atoMEC/writeoutput.py:538]: Haha, well I wanted the headstrings to line up with the number underneath (see e.g. contrasting images below) and didn't immediately know / was too lazy to investigate how to do this the 'proper' way (I guess with string formatting).

I'll look into it unless you know the solution easily? Alternatively, we could just switch to comma separated values (and likewise in the headstring) in which case the lack of alignment wouldn't bother me so much.

output_2
output_1

File: atoMEC/numerov.py (hunk 6 above)
Comment (timcallow) [reply atoMEC/numerov.py:842]: I will do as you suggest with the commit message and remove the commented out code, I agree it isn't so nice. But I'm not sure what you mean about converting the code to a block comment vs adding commented code? I just commented out some old functions.

File: atoMEC/writeoutput.py (hunk 2 above)
Comment (timcallow) [reply atoMEC/writeoutput.py:536]: Ok look at new commit 23c5826 and see what you think

File: atoMEC/writeoutput.py [hunk 7]
```diff
@@ -524,17 +524,15 @@ def density_to_csv(rgrid, density, filename):
...
+        headstr = " ".join(
+            s.rjust(12)
+            for s in [
+                "r (a_0)",
+                "n^u (orbs)",
+                "n^u (ideal)",
+                "n^d (orbs)",
+                "n^d (ideal)",
+            ]
         )
```
Comment (DanielKotik) [inline atoMEC/writeoutput.py:536]: This looks much nicer. I think a comment a la "# right-justifing column heads with a fixed with of 12 characters" would make things even clearer. For now, lets stick with your improved code. Maybe we should add a new issue that explores ways to improve saving of the output, for example by using the `csv` module.

File: atoMEC/writeoutput.py (hunk 2 above)
Comment (DanielKotik) [reply atoMEC/writeoutput.py:536]: See my comment here: https://github.com/atomec-project/atoMEC/pull/123/commits/23c58d6b324599e32f1f3af2b2fcd422c29d50f4#r789684125

**Comment by DanielKotik** [comment]
Please be aware, that some of my suggestions introduced pep8 style issues, see the failed pipeline checks!

File: atoMEC/writeoutput.py (hunk 7 above)
Comment (timcallow) [reply atoMEC/writeoutput.py:537]: Added comment and issue #125 so we don't forget

File: atoMEC/writeoutput.py (hunk 2 above)
Comment (timcallow) [reply atoMEC/writeoutput.py:537]: > Please be aware, that some of my suggestions introduced pep8 style issues, see the failed pipeline checks!

Ah yes all sorted now!

File: atoMEC/numerov.py (hunk 6 above)
Comment (timcallow) [reply atoMEC/numerov.py]: Hmm might be because I was working on two different implementations at the same time, and copied a block of code from the other branch in case it would be useful... Anyway have done as suggested with deletion of commented code and an appropriate commit message
